kong/clustering/init.lua

```diff
@@ -138,205 +39,88 @@ function _M.new(conf)
-  local services_hash  = services  and ngx_md5(to_sorted_string(services))  or DECLARATIVE_EMPTY_CONFIG_HASH
-  local plugins_hash   = plugins   and ngx_md5(to_sorted_string(plugins))   or DECLARATIVE_EMPTY_CONFIG_HASH
-  local upstreams_hash = upstreams and ngx_md5(to_sorted_string(upstreams)) or DECLARATIVE_EMPTY_CONFIG_HASH
-  local targets_hash   = targets   and ngx_md5(to_sorted_string(targets))   or DECLARATIVE_EMPTY_CONFIG_HASH
-
-  config_table.routes    = nil
-  config_table.services  = nil
-  config_table.plugins   = nil
-  config_table.upstreams = nil
-  config_table.targets   = nil
-
-  local config_hash = ngx_md5(to_sorted_string(config_table) .. routes_hash
-                                                             .. services_hash
-                                                             .. plugins_hash
-                                                             .. upstreams_hash
-                                                             .. targets_hash)
-
-  config_table.routes    = routes
-  config_table.services  = services
-  config_table.plugins   = plugins
-  config_table.upstreams = upstreams
-  config_table.targets   = targets
-
-  return config_hash, {
-    config    = config_hash,
-    routes    = routes_hash,
-    services  = services_hash,
-    plugins   = plugins_hash,
-    upstreams = upstreams_hash,
-    targets   = targets_hash,
-  }
+function _M:handle_cp_websocket()
+  return self.json_handler:handle_cp_websocket()
 end
 
-local function fill_empty_hashes(hashes)
-  for _, field_name in ipairs{
-    "config",
-    "routes",
-    "services",
-    "plugins",
-    "upstreams",
-    "targets",
-  } do
-    hashes[field_name] = hashes[field_name] or DECLARATIVE_EMPTY_CONFIG_HASH
-  end
+function _M:handle_wrpc_websocket()
+  return self.wrpc_handler:handle_cp_websocket()
 end
 
-
---- Return the highest supported Hybrid mode protocol version.
-local function check_protocol_support(conf, cert, cert_key)
-  local params = {
-    scheme = "https",
-    method = "HEAD",
-
-    ssl_verify = true,
-    ssl_client_cert = cert,
-    ssl_client_priv_key = cert_key,
-  }
-
-  if conf.cluster_mtls == "shared" then
-    params.ssl_server_name = "kong_clustering"
-
-  else
-    -- server_name will be set to the host if it is not explicitly defined here
-    if conf.cluster_server_name ~= "" then
-      params.ssl_server_name = conf.cluster_server_name
-    end
-  end
-
-  local c = http.new()
-  local res, err = c:request_uri(
-    "https://" .. conf.cluster_control_plane .. "/v1/wrpc", params)
-  if not res then
-    return nil, err
-  end
-
-  if res.status == 404 then
-    return "v0"
-  end
-
-  return "v1"   -- wrpc
+function _M:serve_version_handshake()
+  return version_negotiation.serve_version_handshake(self.conf, self.cert_digest)
 end
 
+function _M:init_cp_worker(plugins_list)
+  self.json_handler.plugins_list = plugins_list
+  self.wrpc_handler.plugins_list = plugins_list
```

Comment:
Why not make `plugins_list` an argument to `init_wokker()` so it is less intrusive.

Reply:
done